### PR TITLE
K8SPXC-529: add graceful operator shutdown

### PR DESCRIPTION
**Pull request by qJkee**
[K8SPXC-529](https://jira.percona.com/browse/K8SPXC-529)

When all objects are being deleted, we should wait for pxc cluster correct deletion first, and only then delete operator

### CODE REVIEW

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| upgrade-haproxy | passed |
| storage | passed |
| haproxy | passed |
| big-data | passed |
| scaling | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| self-healing | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| smart-update | passed |
| restore-to-encrypted-cluster | passed |
| operator-self-healing | failed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| demand-backup | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| upgrade-consistency | passed |
| scheduled-backup | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| scaling | passed |
| upgrade-haproxy | passed |
| haproxy | passed |
| storage | passed |
| recreate | passed |
| init-deploy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| upgrade-proxysql | passed |
| limits | passed |
| security-context | passed |
| monitoring | failed |
| self-healing-advanced | passed |
| smart-update | passed |
| demand-backup | passed |
| monitoring-2-0 | failed |
| operator-self-healing | passed |
| affinity | passed |
| scheduled-backup | passed |
| one-pod | passed |
| auto-tuning | passed |
| upgrade-consistency | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

File: cmd/manager/main.go [hunk 1]
```diff
@@ -127,8 +127,14 @@ func main() {
 
 	log.Info("Starting the Cmd.")
 
+	stopCH, err := k8s.StartStopSignalHandler(mgr.GetClient())
```
Comment (delgod) [inline cmd/manager/main.go:130]: can we pass correct namespace here?
```suggestion
	stopCH, err := k8s.StartStopSignalHandler(mgr.GetClient(), namespace)
```
there are a few cases that should be tested:
1. `WATCH_NAMESPACE` can be equal to an empty string - just pass it as it is to `client.Get`, which should return objects in all namespaces
2. `WATCH_NAMESPACE` can be a comma-separated list `strings.Split(namespace, ",")` - it is better to iterate over it
3. `WATCH_NAMESPACE` can have one namespace name - just pass it as it is to `client.Get`, which should return objects in all namespaces

**Review by delgod — CHANGES_REQUESTED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| storage | passed |
| recreate | passed |
| big-data | passed |
| haproxy | passed |
| upgrade-haproxy | passed |
| init-deploy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring | passed |
| demand-backup | passed |
| smart-update | passed |
| operator-self-healing | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| scheduled-backup | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | failed |

**Review by delgod — APPROVED** [review]

**Comment by JNKPercona** [comment]
PXC operator docker - perconalab/percona-xtradb-cluster-operator:PR-689-20c27128

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| upgrade-haproxy | passed |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-encrypted-with-tls | passed |
| storage | passed |
| self-healing | passed |
| haproxy | passed |
| self-healing-advanced | passed |
| init-deploy | passed |
| operator-self-healing | passed |
| limits | passed |
| scaling | passed |
| monitoring | passed |
| scaling-proxysql | passed |
| monitoring-2-0 | passed |
| security-context | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |

**Review by hors — APPROVED** [review]

File: pkg/k8s/graceful_shutdown.go [hunk 2]
```diff
@@ -0,0 +1,74 @@
+package k8s
+
+import (
+	"context"
+	"time"
+
+	api "github.com/percona/percona-xtradb-cluster-operator/pkg/apis/pxc/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/runtime/signals"
```
Comment (blez) [inline pkg/k8s/graceful_shutdown.go:10]: This package is deprecated

File: pkg/k8s/graceful_shutdown.go [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+)
+
+var (
+	stopCH = make(chan struct{}, 0)
```
Comment (blez) [inline pkg/k8s/graceful_shutdown.go:14]: 1. 0 is not needed
2. `stopCH` can be local var instead of global

File: pkg/k8s/graceful_shutdown.go [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+	close(stopCH)
+}
+
+// Stop is used to understand, when we need to stop operator(usially SIGTERM)
+// to start cleanup process and delete required pxc clusters in current(operator)
+// namespace. See K8SPXC-529
+func stop(cl client.Client, namespaces []string) {
+	log.Info("Got stop signal, starting to list clusters")
+
+	for {
```
Comment (blez) [inline pkg/k8s/graceful_shutdown.go:38]: Should we have some timeout here? Not sure but I assume that loop may be infinite.

File: pkg/k8s/graceful_shutdown.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+				}
+			}
+			if clusterDeleteInProgress == 0 {
+				inProgress--
+			}
+		}
+
+		if inProgress == 0 {
+			return
+		}
```
Comment (blez) [inline pkg/k8s/graceful_shutdown.go:72]: You do not need those `clusterDeleteInProgress` and `inProgress`. Current implementation loop over all clusters in all namespaces. But as I understand this logic, we can stop doing iteration when we find the first cluster where  `len(v.Finalizers) > 0`
So just move this logic to separate func and 
```
if len(v.Finalizers) > 0 {
	return false
}
```

**Review by blez — CHANGES_REQUESTED** [review]

File: pkg/k8s/graceful_shutdown.go (hunk 4 above)
Comment (qJkee) [reply pkg/k8s/graceful_shutdown.go:38]: pod will be force killed after timeout, which is configurable from yaml files

File: pkg/k8s/graceful_shutdown.go (hunk 3 above)
Comment (qJkee) [reply pkg/k8s/graceful_shutdown.go]: 3d14adf9

File: pkg/k8s/graceful_shutdown.go (hunk 2 above)
Comment (qJkee) [reply pkg/k8s/graceful_shutdown.go]: 4ebf23dc

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| upgrade-haproxy | passed |
| recreate | passed |
| storage | passed |
| big-data | passed |
| haproxy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| init-deploy | passed |
| restore-to-encrypted-cluster | failed |
| security-context | failed |
| upgrade-proxysql | passed |
| limits | passed |
| monitoring | failed |
| monitoring-2-0 | failed |
| demand-backup | failed |
| affinity | passed |
| scheduled-backup | failed |
| demand-backup-encrypted-with-tls | failed |
| smart-update | passed |
| one-pod | failed |
| self-healing-advanced | passed |
| auto-tuning | passed |
| operator-self-healing | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Review by delgod — DISMISSED** [review]

File: pkg/k8s/graceful_shutdown.go (hunk 5 above)
Comment (qJkee) [reply pkg/k8s/graceful_shutdown.go]: 59a49217

**Review by delgod — DISMISSED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| storage | passed |
| haproxy | passed |
| big-data | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| security-context | failed |
| restore-to-encrypted-cluster | failed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring | failed |
| monitoring-2-0 | failed |
| affinity | passed |
| smart-update | failed |
| demand-backup | failed |
| scheduled-backup | failed |
| one-pod | failed |
| demand-backup-encrypted-with-tls | failed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| operator-self-healing | passed |
| tls-issue-self | passed |
| upgrade-consistency | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| big-data | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| recreate | passed |
| storage | passed |
| haproxy | passed |
| init-deploy | passed |
| self-healing | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| self-healing-advanced | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| smart-update | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| operator-self-healing | passed |
| scheduled-backup | passed |
| users | failed |
| upgrade-consistency | passed |
| demand-backup-encrypted-with-tls | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

**Comment by JNKPercona** [comment]
PXC operator docker - perconalab/percona-xtradb-cluster-operator:PR-689-27caec8b

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| recreate | passed |
| restore-to-encrypted-cluster | passed |
| demand-backup | passed |
| scheduled-backup | passed |
| demand-backup-encrypted-with-tls | passed |
| upgrade-haproxy | passed |
| scaling | passed |
| scaling-proxysql | passed |
| security-context | passed |
| storage | passed |
| self-healing | passed |
| self-healing-advanced | passed |
| big-data | passed |
| operator-self-healing | passed |
| haproxy | passed |
| init-deploy | passed |
| limits | passed |
| monitoring-2-0 | passed |
| affinity | passed |
| one-pod | passed |
| auto-tuning | passed |
| proxysql-sidecar-res-limits | passed |
| users | passed |
| upgrade-proxysql | passed |
| smart-update | passed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |

File: pkg/k8s/graceful_shutdown.go [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+		for _, ns := range namespaces {
+
+			clusterList := &api.PerconaXtraDBClusterList{}
+
+			err := cl.List(context.TODO(), clusterList, &client.ListOptions{
+				Namespace: ns,
+			})
+			if err != nil {
+				log.Error(err, "list clusters in namespace", "namespace", ns)
+				continue
```
Comment (blez) [inline pkg/k8s/graceful_shutdown.go:49]: Imagine we have namespace1,  namespace2 and namespace3.
For namespace 1 and 2 `List` operation ends with error. But for namespace3 it ends successfully.
Then namespace1 and 2 just ignored.

Comment (delgod) [reply pkg/k8s/graceful_shutdown.go:49]: what is the expected behavior?
Should we lock deleting if we cannot list clusters in namespace1 and namespace2?

Comment (qJkee) [reply pkg/k8s/graceful_shutdown.go:49]: 7559c1ef

**Review by blez — APPROVED** [review]

**Comment by JNKPercona** [comment]
| Test name  | Status |
| ------------- | ------------- |
| scaling | passed |
| storage | passed |
| recreate | passed |
| haproxy | passed |
| upgrade-haproxy | passed |
| big-data | passed |
| self-healing | passed |
| init-deploy | passed |
| scaling-proxysql | passed |
| restore-to-encrypted-cluster | passed |
| security-context | passed |
| upgrade-proxysql | passed |
| limits | passed |
| monitoring-2-0 | passed |
| demand-backup | passed |
| smart-update | passed |
| affinity | passed |
| self-healing-advanced | passed |
| one-pod | passed |
| auto-tuning | passed |
| operator-self-healing | passed |
| proxysql-sidecar-res-limits | failed |
| scheduled-backup | passed |
| users | failed |
| upgrade-consistency | passed |
| tls-issue-self | passed |
| tls-issue-cert-manager | passed |
| tls-issue-cert-manager-ref | passed |
| validation-hook | passed |
| demand-backup-encrypted-with-tls | failed |

**Review by delgod — APPROVED** [review]

**Pull request merged**